crates/apollo-compiler/src/database/hir.rs

```diff
@@ -1288,6 +1288,92 @@ impl Float {
     }
 }
 
+/// When some item (for example fields of an object type) have unique names
+/// and may be found on type extensions, this pre-computes where to find
+/// each item from its name
+#[derive(Clone, Debug, Eq)]
+pub(crate) struct ByNameWithExtensions {
+    /// `(None, i)` designates `def.example[i]`.
+    /// `(Some(j), i)` designates `def.extensions[j].example[i]`.
+    indices: IndexMap<String, (Option<usize>, usize)>,
+}
+
+/// Equivalent to ignoring a `ByNameWithExtensions` field in `PartialEq` for its parent struct,
+/// since it is determined by other fields.
+impl PartialEq for ByNameWithExtensions {
+    fn eq(&self, _other: &Self) -> bool {
+        true
+    }
+}
+
+/// Equivalent to ignoring a `ByNameWithExtensions` field in `Hash` for its parent struct,
+/// since it is determined by other fields.
+impl hash::Hash for ByNameWithExtensions {
+    fn hash<H: hash::Hasher>(&self, _state: &mut H) {
+        // do nothing
+    }
+}
+
+impl ByNameWithExtensions {
+    pub(crate) fn new<Item>(self_items: &[Item], name: impl Fn(&Item) -> &str) -> Self {
+        let mut indices = IndexMap::new();
+        for (i, item) in self_items.iter().enumerate() {
+            indices.entry(name(item).to_owned()).or_insert((None, i));
+        }
+        ByNameWithExtensions { indices }
+    }
+
+    pub(crate) fn add_extension<Item>(
+        &mut self,
+        extension_index: usize,
+        extension_items: &[Item],
+        name: impl Fn(&Item) -> &str,
+    ) {
+        for (i, item) in extension_items.iter().enumerate() {
+            self.indices
+                .entry(name(item).to_owned())
+                .or_insert((Some(extension_index), i));
+        }
+    }
```

Comment:
so this is perhaps dependent on how accurate we want to be in our HIR. An extension should not exist if the original definition doesn't exist. Perhaps this method should also not add an extension if we don't have an existing, matching definition. 

it'd have to check if the definition already exists, and if the definition is of the same type as the proposed extension. Much like quite a few of our recent conversations - our HIR creation is doing a bunch of the same work as validation. This would be adding even more of that work to extensions.

Reply:
I don’t understand what you mean. This is a method of `ByNameWithExtensions` which is used in the `ObjectTypeDefinition` struct. If a object type extension extends an object type that doesn’t exist, then its `ByNameWithExtensions` also doesn’t exist and this code isn’t called. This happens if `get_mut` here returns `None`:

https://github.com/apollographql/apollo-rs/blob/3088432300072926635df95072963816f52cdf7b/crates/apollo-compiler/src/database/hir_db.rs#L384